### PR TITLE
BUG: merge should upcast to highest resolution

### LINKED ISSUE #55212: BUG:
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
t1 = '2023-09-01'
t2 = '2023-09-01 01:00:00'
t3 = '2023-09-01 01:30:00'
ds1 = pd.date_range(t1, t2, freq='30T')
ds2 = pd.date_range(t1, t3, freq='30T')
df1 = pd.DataFrame({
    'ds': ds1.astype('datetime64[us]'),
    'y1': range(len(ds1)),
})
df2 = pd.DataFrame({
    'ds': ds2.astype('datetime64[ns]'),
    'y2': range(len(ds2)),
})
df3 = df1.merge(df2, on=['ds'], how='outer') #will only convert df2 `ds` type to df1 `ds` type
df3.dtypes
```

### Issue Description

The column type was not converted from the lower resolution (dtype('<M8[us]')) to the higher resolution (dtype('<M8[ns]')). 

It only converted the type in df2 to the type in df2 for the `on` columns. 
```
File ~\conda-envs\dev\lib\site-packages\pandas\core\arrays\_mixins.py:398, in NDArrayBackedExtensionArray._putmask(self, mask, value)
    383 """
    384 Analogue to np.putmask(self, mask, value)
    385 
   (...)
    394     If value cannot be cast to self.dtype.
    395 """
    396 value = self._validate_setitem_value(value)
--> 398 np.putmask(self._ndarray, mask, value)

File <__array_function__ internals>:180, in putmask(*args, **kwargs)

TypeError: Cannot cast array data from dtype('<M8[ns]') to dtype('<M8[us]') according to the rule 'safe'
```

### Expected Behavior

I expect `df1.merge(df2, on=['ds'], how='outer')` and `df2.merge(df1, on=['ds'], how='outer')` both should work in this case.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : ba1cccd19da778f0c3a7d6a885685da16a072870
python              : 3.9.15.final.0
python-bits         : 64
OS                  : Windows
OS-release          : 10
Version             : 10.0.19044
machine             : AMD64
processor           : Intel64 Family 6 Model 58 Stepping 0, GenuineIntel
byteorder           : little
LC_ALL              : None
LANG                : None
LOCALE              : English_Australia.1252

pandas              : 2.1.0
numpy               : 1.23.3
pytz                : 2022.1
dateutil            : 2.8.2
setuptools          : 65.5.0
pip                 : 22.3.1
Cython              : 0.29.30
pytest              : 7.1.2
hypothesis          : None
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : 4.9.0
html5lib            : None
pymysql             : None
psycopg2            : 2.9.3
jinja2              : 3.0.3
IPython             : 8.4.0
pandas_datareader   : None
bs4                 : 4.11.1
bottleneck          : 1.3.5
dataframe-api-compat: None
fastparquet         : None
fsspec              : 2022.5.0
gcsfs               : None
matplotlib          : 3.5.2
numba               : 0.53.0
numexpr             : 2.7.3
odfpy               : None
openpyxl            : 3.0.9
pandas_gbq          : None
pyarrow             : 12.0.0
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : 1.8.1
sqlalchemy          : 1.4.46
tables              : None
tabulate            : None
xarray              : 2022.3.0
xlrd                : None
zstandard           : None
tzdata              : 2022.1
qtpy                : None
pyqt5               : None

</details>

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #55212  (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by lithomas1** [comment]
@jbrockmendel 

This only gets some of the cases.

I think this misses when we take everything from the left join col.
Any idea where to handle the upcasting then?

File: doc/source/whatsnew/v2.1.4.rst [hunk 1]
```diff
@@ -26,6 +26,7 @@ Bug fixes
 - Fixed bug in :func:`to_numeric` converting to extension dtype for ``string[pyarrow_numpy]`` dtype (:issue:`56179`)
 - Fixed bug in :meth:`DataFrame.__setitem__` casting :class:`Index` with object-dtype to PyArrow backed strings when ``infer_string`` option is set (:issue:`55638`)
 - Fixed bug in :meth:`DataFrame.to_hdf` raising when columns have ``StringDtype`` (:issue:`55088`)
+- Fixed bug in :meth:`DataFrame.merge` not being able to join on ``datetime64`` columns or differing resolutions (:issue:`55212`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.1.4.rst:29]: or->of

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -1073,16 +1073,8 @@ def _maybe_add_join_keys(
...
                     result_dtype = find_common_type([lvals.dtype, rvals.dtype])
-                    if (
-                        lvals.dtype.kind == "M"
-                        and rvals.dtype.kind == "M"
-                        and result_dtype.kind == "O"
-                    ):
-                        # TODO(non-nano) Workaround for common_type not dealing
-                        # with different resolutions
-                        result_dtype = key_col.dtype
+                    key_col = Index(lvals).astype(result_dtype).where(~mask_left, rvals)
```
Comment (jbrockmendel) [inline pandas/core/reshape/merge.py:1077]: pass copy=False?

**Comment by jbrockmendel** [comment]
> I think this misses when we take everything from the left join col.  Any idea where to handle the upcasting then?

Do you have an example?  That'd save me some time in figuring out what code paths that case goes through

**Comment by lithomas1** [comment]
Sorry for the slow reply.

I think my changes here only fixed the result for outer/right merges. The tests that I've added that are failing are doing so for inner/left merges.

Here's a MRE.
```
t1 = '2023-09-01'
t2 = '2023-09-01 01:00:00'
t3 = '2023-09-01 01:30:00'
ds1 = pd.date_range(t1, t2, freq='30T')
ds2 = pd.date_range(t1, t3, freq='30T')
df1 = pd.DataFrame({
    'ds': ds1.astype('datetime64[us]'),
    'y1': range(len(ds1)),
})
df2 = pd.DataFrame({
    'ds': ds2.astype('datetime64[ns]'),
    'y2': range(len(ds2)),
})
# Changing this to outer/right makes it work
df3 = df1.merge(df2, on=['ds'], how='left') #will only convert df2 `ds` type to df1 `ds` type
df3.dtypes
```

File: doc/source/whatsnew/v2.2.0.rst [hunk 3]
```diff
@@ -761,6 +761,7 @@ Datetimelike
 - Bug in parsing datetime strings with nanosecond resolution with non-ISO8601 formats incorrectly truncating sub-microsecond components (:issue:`56051`)
 - Bug in parsing datetime strings with sub-second resolution and trailing zeros incorrectly inferring second or millisecond resolution (:issue:`55737`)
 - Bug in the results of :func:`to_datetime` with an floating-dtype argument with ``unit`` not matching the pointwise results of :class:`Timestamp` (:issue:`56037`)
+- Fixed bug in :meth:`DataFrame.merge` not being able to join on ``datetime64`` columns of differing resolutions (:issue:`55212`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.2.0.rst:764]: in main they are able to join just gives the wrong result reso?

**Comment by jbrockmendel** [comment]
The code change looks good to me.  The test change looks plausible but i haven't examined closely.

**Comment by mroeschke** [comment]
Looks like this PR has gone stale and needs a rebase. Closing to clear the queue but feel free to reopen when you have time to circle back

**Pull request closed without merging**